reactive/kubernetes_master.py

```diff
@@ -2418,3 +2437,96 @@ def send_registry_location():
 @when('config.changed.image-registry')
 def send_new_registry_location():
     clear_flag('kubernetes-master.sent-registry')
+
+
+@when('leadership.is_leader',
+      'leadership.set.kubernetes-master-addons-restart-for-ca',
+      'kubernetes-master.components.started')
+def restart_addons_for_ca():
+    try:
+        # Get deployments/daemonsets/statefulsets
+        cmd = [
+            'kubectl', 'get',
+            '-o', 'json',
+            '--all-namespaces',
+            '-l', 'cdk-restart-on-ca-change=true',
+            'daemonset,deployment,statefulset'
+        ]
+        output = check_output(cmd).decode('UTF-8')
+        deployments = json.loads(output)['items']
+
+        # Get ServiceAccounts
+        service_account_names = set(
+            (
+                deployment['metadata']['namespace'],
+                deployment['spec']['template']['spec'].get(
+                    'serviceAccountName', 'default'
+                )
+            )
+            for deployment in deployments
+        )
+        service_accounts = []
+        for namespace, name in service_account_names:
+            cmd = [
+                'kubectl', 'get',
+                '-o', 'json',
+                '-n', namespace,
+                'ServiceAccount', name
+            ]
+            output = check_output(cmd).decode('UTF-8')
+            service_account = json.loads(output)
+            service_accounts.append(service_account)
+
+        # Get ServiceAccount secrets
+        secret_names = set()
+        for service_account in service_accounts:
+            namespace = service_account['metadata']['namespace']
+            for secret in service_account['secrets']:
+                secret_names.add((namespace, secret['name']))
+        secrets = []
+        for namespace, name in secret_names:
+            cmd = [
+                'kubectl', 'get',
+                '-o', 'json',
+                '-n', namespace,
+                'Secret', name
+            ]
+            output = check_output(cmd).decode('UTF-8')
+            secret = json.loads(output)
+            secrets.append(secret)
+
+        # Check secrets have updated CA
+        with open(ca_crt_path, 'rb') as f:
+            ca = f.read()
+        encoded_ca = base64.b64encode(ca).decode('UTF-8')
+        mismatched_secrets = [
+            secret for secret in secrets
+            if secret['data']['ca.crt'] != encoded_ca
```

Comment:
Will this trip up if there are secrets named ca.cert in the cluster that aren't meant to be the same as this encoded ca?

Reply:
Nevermind, I see the selector on the query. We'll just need to be careful not to tag something that uses a different ca.